cirq/protocols/equal_up_to_global_phase.py

```diff
@@ -0,0 +1,171 @@
+        the differences in _magnitudes_ of two compared complex numbers.
+
+    Returns:
+        True if objects are approximately equal up to phase, False otherwise.
+    """
+
+    # fall back to _equal_up_to_global_phase_ for val.
+    eq_up_to_phase_getter = getattr(val, '_equal_up_to_global_phase_', None)
+    if eq_up_to_phase_getter is not None:
+        result = eq_up_to_phase_getter(other, atol)
+        if result is not NotImplemented:
+            return result
+
+    # fall back to _equal_up_to_global_phase_ for other.
+    other_eq_up_to_phase_getter = getattr(
+        other, '_equal_up_to_global_phase_', None
+    )
+    if other_eq_up_to_phase_getter is not None:
+        result = other_eq_up_to_phase_getter(val, atol)
+        if result is not NotImplemented:
+            return result
+
+    # Try to compare the magnitude of two complex numbers.
+    # FIXME: why restrict casting int, float -> complex for comparison?
+    if isinstance(val, complex):
+        if not isinstance(other, complex):
+            return False
+        return cirq.approx_eq(np.abs(val), np.abs(other), atol=atol)
+
+    # Try to compare source and target recursively, assuming they're iterable.
+    result = _eq_up_to_phase_iterables(val, other, atol=atol)
+
+    # Fallback to cir approx_eq for remaining types
+    if result is NotImplemented:
+        return cirq.approx_eq(val, other, atol=atol)
+    return result
+
+
+def _eq_up_to_phase_iterables(
+    val: Any,
+    other: Any,
+    *,
+    atol: Union[int, float]
+) -> bool:
+    """Iterates over arguments and checks for equality up to global phase.
+
+    For iterables of the same length and comparable structure, check that the
+    difference between phases of corresponding elements can be described by a
+    single complex value.
+
+    Args:
+        val: Source for comparison.
+        other: Target for comparison.
+        atol: The minimum absolute tolerance.
+
+    Returns:
+        True if objects are approximately equal up to phase, False otherwise.
+        Returns NotImplemented when approximate equality is not implemented
+        for given types.
+    """
+
+    if not (isinstance(val, Iterable) and isinstance(other, Iterable)):
+        return NotImplemented
+
+    val_it = iter(val)
+    other_it = iter(other)
+    global_phase = None
+
+    while True:
+        try:
+            val_next = next(val_it)
+        # only allow phase comparison for equal-length containers
+        except StopIteration:
+            try:
+                next(other_it)
+                return False
+            except StopIteration:
+                return True
+
+        try:
+            other_next = next(other_it)
+        except StopIteration:
+            return False
+        # check that a single value describes all phase differences between
+        # corresponding values in the containers
+        if global_phase is None:
+            global_phase = _phase_difference(val_next, other_next)

```

Comment:
This check is not correct, because it will fail when the first two elements are zero (and so appear to have the same phase even if the two sequences don't).

I think we should avoid handling arbitrary iterables, because it's not clear to me whether users will think that the "up to phase" means per-pair or across-whole-array. We could potentially support numpy arrays, because I think in that case it is clear what is meant and also we have existing code this case in linalg.

Reply:
![image](https://user-images.githubusercontent.com/14928221/58579565-fbf90780-8210-11e9-97c1-c75ab31df2c8.png)

^ not sure what this "mypy" test is complaining about or how to address it.